src/number_theory/well_approximable.lean

```diff
@@ -0,0 +1,313 @@
+  a ∈ approx_order_of A n δ ↔ ∃ (b : A), order_of b = n ∧ a ∈ ball b δ :=
+by simp only [approx_order_of, thickening_eq_bUnion_ball, mem_Union₂, mem_set_of_eq, exists_prop]
+
+/-- In a seminormed group `A`, given a sequence of distances `δ₁, δ₂, ...`, `well_approximable A δ`
+is the limsup as `n → ∞` of the sets `approx_order_of A n δₙ`. Thus, it is the set of points that
+lie in infinitely many of the sets `approx_order_of A n δₙ`. -/
+@[to_additive add_well_approximable "In a seminormed additive group `A`, given a sequence of
+distances `δ₁, δ₂, ...`, `add_well_approximable A δ` is the limsup as `n → ∞` of the sets
+`approx_add_order_of A n δₙ`. Thus, it is the set of points that lie in infinitely many of the sets
+`approx_add_order_of A n δₙ`."]
+def well_approximable (A : Type*) [seminormed_group A] (δ : ℕ → ℝ) : set A :=
+blimsup (λ n, approx_order_of A n (δ n)) at_top (λ n, 0 < n)
+
+@[to_additive mem_add_well_approximable_iff]
+lemma mem_well_approximable_iff {A : Type*} [seminormed_group A] {δ : ℕ → ℝ} {a : A} :
+  a ∈ well_approximable A δ ↔ a ∈ blimsup (λ n, approx_order_of A n (δ n)) at_top (λ n, 0 < n) :=
+iff.rfl
+
+namespace approx_order_of
+
+variables {A : Type*} [seminormed_comm_group A] {a : A} {m n : ℕ} (δ : ℝ)
+
+@[to_additive]
+lemma image_pow_subset_of_coprime (hm : 0 < m) (hmn : n.coprime m) :
+  (λ y, y^m) '' (approx_order_of A n δ) ⊆ approx_order_of A n (m * δ) :=
+begin
+  rintros - ⟨a, ha, rfl⟩,
+  obtain ⟨b, hb, hab⟩ := mem_approx_order_of_iff.mp ha,
+  replace hb : b^m ∈ {u : A | order_of u = n}, { rw ← hb at hmn ⊢, exact order_of_pow_coprime hmn },
+  apply ball_subset_thickening hb ((m : ℝ) • δ),
+  convert pow_mem_ball hm hab using 1,
+  simp only [nsmul_eq_mul, algebra.id.smul_eq_mul],
+end
+
+@[to_additive]
+lemma image_pow_subset (n : ℕ) (hm : 0 < m) :
+  (λ y, y^m) '' (approx_order_of A (n * m) δ) ⊆ approx_order_of A n (m * δ) :=
+begin
+  rintros - ⟨a, ha, rfl⟩,
+  obtain ⟨b, hb : order_of b = n * m, hab : a ∈ ball b δ⟩ := mem_approx_order_of_iff.mp ha,
+  replace hb : b^m ∈ {y : A | order_of y = n},
+  { rw [mem_set_of_eq, order_of_pow' b hm.ne', hb, nat.gcd_mul_left_left, n.mul_div_cancel hm], },
+  apply ball_subset_thickening hb (m * δ),
+  convert pow_mem_ball hm hab,
+  simp only [nsmul_eq_mul],
+end
+
+@[to_additive]
+lemma smul_subset_of_coprime (han : (order_of a).coprime n) :
+  a • approx_order_of A n δ ⊆ approx_order_of A ((order_of a) * n) δ :=
+begin
+  simp_rw [approx_order_of, thickening_eq_bUnion_ball, ← image_smul, image_Union₂,
+    image_smul, smul_ball'', smul_eq_mul, mem_set_of_eq],
+  refine Union₂_subset_iff.mpr (λ b hb c hc, _),
+  simp only [mem_Union, exists_prop],
+  refine ⟨a * b, _, hc⟩,
+  rw ← hb at ⊢ han,
+  exact (commute.all a b).order_of_mul_eq_mul_order_of_of_coprime han,
+end
+
+@[to_additive vadd_eq_of_mul_dvd]
+lemma smul_eq_of_mul_dvd (hn : 0 < n) (han : (order_of a) * (order_of a) ∣ n) :
```

Comment:
Is it really easier to use with this weird spelling of `(order_of a)^2`? On paper I guess you would never write this.

Reply:
I've switched to use `(order_of a)^2`.

Using `^2` introduces some pointless indirection but your comment persuades me this is the right compromise.